Add collections module

ADBeveridge opened the pull request.

ADBeveridge committed:
ospl: add collections module

Signed-off-by: Alan Beveridge <adgbeveridge@proton.me>

Sulfiore commented on src/collection.rs:
```diff
@@ -0,0 +1,127 @@
+	fn insert_into (&self, fs: &Filesystem) -> Result<(), Error> {
+		Ok(())
+	}
+	#[allow(unused_variables)]
```
Please space out your code one empty line after each function is great

Sulfiore commented on src/collection.rs:
```diff
@@ -0,0 +1,127 @@
+
+	/*pub fn from_folder(&mut self, _db: &Database, collection_path: &str) -> Result <(), Error>
+	{
+		// TODO: Implement.
```
I suggest you replace this whole comment section with only a TODO that explains what to implement, because the code might not look like this while we come there.

Sulfiore commented on src/element.rs:
```diff
@@ -37,5 +37,6 @@ pub trait ElementDatabase
 pub trait ElementFilesystem
 {
 	fn insert_into(&self, fs: &Filesystem) -> Result<(), Error>;
+	fn create(&mut self, fs: &Filesystem) -> Result<(), Error>;
```
I would suggest if something doesn't apply to every element, we don't put it into the Element trait. Or we create a new trait that apply only to containers (Collections, albums)

Sulfiore commented on src/collection.rs:
```diff
@@ -0,0 +1,127 @@
+	}*/
+}
+
+impl Collection
```
Why is this empty ?

ADBeveridge replied:
Just a hangover from the photo.rs file. I got rid of it.

Sulfiore commented on src/photo.rs:
```diff
@@ -154,6 +154,10 @@ impl ElementFilesystem for Photo
 		Ok(())
 	}
+	#[allow(unused_variables)]
+	fn create(&mut self, fs: &Filesystem) -> Result<(), Error> {
```
Due to my prior comment on element.rs, it would be best to remove this.

Sulfiore commented on src/filesystem.rs:
```diff
@@ -89,4 +89,9 @@ impl Filesystem
-}
+	
+	pub(crate) fn create_file(&self, object: &mut dyn ElementFilesystem) -> Result<(), Error>
+	{
```
I'm not sure that it is necessary to do this, Couldn't the `insert` method be a more appropriate solution for this task?

Sulfiore commented on src/database.rs:
```diff
@@ -57,7 +57,10 @@ impl Database
-			Ok(_) => { return Ok(db); },
+			Ok(_) =>
+			{
+				return Ok(db);
```
I changed this in the code due to an issue with the code coverage not counting this line when it is properly indented. Can you please revert it back to its original state ?

Sulfiore commented on src/directory.rs:
```diff
@@ -54,4 +54,4 @@ impl Directory
 		Ok(())
 	}
-}
+}
```
It would likely be more appropriate to include this as a separate commit focused on cosmetic changes.

Sulfiore commented on src/photo.rs:
```diff
@@ -181,4 +185,4 @@ fn get_filename_from<P: AsRef<Path>>(path: P) -> String
 	.unwrap()
 	.to_string()
-}
+}
```
cf. other comment on cosmetic changes

ADBeveridge committed:
tests: rename library tests file and fix tmpdir path

 - Rename basic_tests.rs to library_tests.rs
 - Use Cargo environment variable instead of /tmp/ for system independence

Signed-off-by: Alan Beveridge <adgbeveridge@proton.me>
